rocAL/rocAL/source/readers/image/external_source_reader.cpp

```diff
@@ -0,0 +1,261 @@
+      height = std::get<3>(_file_data[cur_idx]);
+      channels = std::get<4>(_file_data[cur_idx]);
+    }
+}
+
+int ExternalSourceReader::close()
+{
+    return release();
+}
+
+ExternalSourceReader::~ExternalSourceReader()
+{
+    release();
+}
+
+int
+ExternalSourceReader::release()
+{
+    if (_mode != FileMode::FILENAME) {
+        if(!_current_fPtr)
+            return 0;
+        fclose(_current_fPtr);
+        _current_fPtr = nullptr;
+        _end_of_sequence = false;   // reset for looping
+    }
+    return 0;
+}
+
+
+void ExternalSourceReader::reset()
+{
+    if (_shuffle) std::random_shuffle(_file_data.begin(), _file_data.end());
+    _read_counter = 0;
+    _curr_file_idx = 0;
+    _end_of_sequence = false;   // reset for looping
+}
+
+size_t ExternalSourceReader::get_file_shard_id()
+{
+    if(_batch_count == 0 || _shard_count == 0)
+        THROW("Shard (Batch) size cannot be set to 0")
+    return _file_id  % _shard_count;
+}
+
+
+void ExternalSourceReader::push_file_name(const std::string& file_name)
+{
+    std::unique_lock<std::mutex> lock(_lock);
+    _file_names_que.push(file_name);
+    lock.unlock();
+    // notify waiting thread of new data
+    _wait_for_input.notify_all();
+}
+
+bool ExternalSourceReader::pop_file_name(std::string& file_name)
+{
+    std::unique_lock<std::mutex> lock(_lock);
+    if(_file_names_que.empty() && !_end_of_sequence)
+        _wait_for_input.wait(lock);
+    if (!_file_names_que.empty()) {
+      file_name = _file_names_que.front();
+      _file_names_que.pop();
+      return true;
+    }else
+      return false;
+
+}
+
+void ExternalSourceReader::push_file_data(std::tuple<unsigned char*, size_t, int, int, int>& image)
+{
+    std::unique_lock<std::mutex> lock(_lock);
+    _images_data_q.push(image);
+    lock.unlock();
+    // notify waiting thread of new data
+    _wait_for_input.notify_all();
+}
+
+bool ExternalSourceReader::pop_file_data(std::tuple<unsigned char*,  size_t, int, int, int>& image)
+{
+    std::unique_lock<std::mutex> lock(_lock);
+    if(_images_data_q.empty() && !_end_of_sequence)
+        _wait_for_input.wait(lock);
+    if (!_images_data_q.empty()) {
+      image = _images_data_q.front();
+      _images_data_q.pop();
+      return true;
+    }else
+      return false;
+}
+
+
+
+void ExternalSourceReader::feed_file_names(const std::vector<std::string>& file_names, size_t num_images, bool eos)
+{
+    for (unsigned n = 0; n < num_images; n++) {
+      push_file_name(file_names[n]);
+    }
+    _end_of_sequence = eos;
+}
+
+void ExternalSourceReader::feed_data(const std::vector<unsigned char *>& images, const std::vector<size_t>& image_size, int mode, bool eos, int width, int height, int channels)
+{
+    for (unsigned n = 0; n < images.size(); n++) {
+        std::tuple<unsigned char*, size_t, int, int, int> image =  std::make_tuple(images[n], image_size[n], width, height, channels);
+        push_file_data(image);
+    }
+    _end_of_sequence = eos;
+}
+
```

Comment:
Please remove blank lines at end of file
There are 4 blank lines, remove 3 of them